### DISCUSSION TITLE
Help to understand the tpm logs

### DISCUSSION THREAD

**Discussion by ayankuma** [stefanberger/swtpm#692] [General]
Hi,

I am running swtpm in one terminal and qemu in another terminal
```
/Users/ayankuma/alpine_vm/swtpm # swtpm socket --tpmstate dir=/tmp/myvtpm2 --ctrl type=unixio,path=/tmp/myvtpm2/swtpm-sock --log level=20 --tpm2
 Ctrl Cmd: length 4
 00 00 00 10 
 Ctrl Rsp: length 4
 00 00 00 00 
 SWTPM_IO_Read: length 10
 80 01 00 00 00 0A 00 00 01 81 
 SWTPM_IO_Write: length 10
 80 01 00 00 00 0A 00 00 01 01 
 Ctrl Cmd: length 4
 00 00 00 01 
 Ctrl Rsp: length 8
 00 00 00 00 00 00 FF FF 
 Ctrl Cmd: length 4
 00 00 00 0E 
 Ctrl Rsp: length 4
 00 00 00 00 
 Ctrl Cmd: length 8
 00 00 00 11 00 00 00 00 
 Ctrl Rsp: length 16
 00 00 00 00 00 00 10 00 00 00 0A F8 00 00 10 00 
 Ctrl Cmd: length 4
 00 00 00 0E 
 Ctrl Rsp: length 4
 00 00 00 00 
 Ctrl Cmd: length 8
 00 00 00 11 00 00 10 00 
 Ctrl Rsp: length 16
 00 00 00 00 00 00 10 00 00 00 0A F8 00 00 10 00 
 Ctrl Cmd: length 8
 00 00 00 02 00 00 00 00 
 Ctrl Rsp: length 4
 00 00 00 00 
```

1. I see the above logs. Can you provide me any pointers to interpret these commands ?
2. I do not see any tpm/tis logs in dmesg.  I have "CONFIG_TCG_TPM=y" in my kernel config. My dts looks like below
```
        platform@c000000 {   
                #address-cells = <0x01>;      
                #size-cells = <0x01>;      
                interrupt-parent = <0x8005>;      
                compatible = "qemu,platform\0simple-bus";
                ranges = <0x00 0x00 0xc000000 0x2000000>;
                             
                tpm_tis@0 {  
                        compatible = "tcg,tpm-tis-mmio";
                        reg = <0x00 0x5000>;
                };                                
        };
```
And my qemu command line is as below. Can you let me know what I am missing ?
```
./build/qemu-system-aarch64 -cpu cortex-a53 -M virt -smp 4 -m 4096 -nographic -chardev   socket,id=chrtpm,path=/tmp/myvtpm2/swtpm-sock -tpmdev   emulator,id=tpm0,chardev=chrtpm -device tpm-tis-device,tpmdev=tpm0 -kernel /Users/ayankuma/alpine_vm/linux/arch/arm64/boot/Image
```

**Comment by stefanberger** [comment]
Control channel commands are defined here: https://github.com/stefanberger/swtpm/blob/master/man/man3/swtpm_ioctls.pod

The values for the control channel commands are defined here: https://github.com/stefanberger/swtpm/blob/v0.7.2/include/swtpm/tpm_ioctl.h#L284-L303

When they are *actual* ioctls then theses ones here are used: https://github.com/stefanberger/swtpm/blob/v0.7.2/include/swtpm/tpm_ioctl.h#L253-L272

The TPM commands are defined in TCG specs for example here ('Commands'): https://trustedcomputinggroup.org/resource/tpm-library-specification/

**Comment by stefanberger** [comment]
As for aarch64: I am afraid I cannot help much there. The interactions between QEMU and swtpm looked good to me from what I can remember. The command line I used is described here: https://github.com/stefanberger/swtpm/issues/493#issuecomment-885221109

**Comment by ayankuma** [comment]
Many thanks for your inputs. 
By adding logs in qemu/backends/tpm/tpm_emulator.c and swtpm/src/swtpm/ctrlchannel.c, I could see that the commands are CMD_SET_DATAFD, CMD_GET_CAPABILITY, CMD_STOP, CMD_SET_BUFFERSIZE, CMD_INIT, CMD_GET_TPMESTABLISHED and CMD_SET_LOCALITY.

With the command line and fedora images that you had used, I could see tpm getting probed 

```
[root@localhost-live ~]# dmesg | grep -i tpm
[    0.000000] efi: SMBIOS 3.0=0x13f560000 TPMFinalLog=0x13c030000 MEMATTR=0x13cf74018 TPMEventLog=0x13b957018 MEMRESERVE=0x13c1d1118 
[    9.918979] tpm_tis c000000.tpm_tis: 2.0 TPM (device-id 0x1, rev-id 1)
```

Are there are sample applications for linux which uses swtpm. At this point, I am trying to determine if the interaction between an application (using /dev/tpm0) <---> linux <---> qemu <---> swtpm works fine or no.

**Reply by stefanberger** [reply]
You mean you want to run a sample application inside the VM that shows that swtpm connected to the VM for TPM emulation purposes is being used?

Inside the VM:
```
# dnf -y install tpm2-tools
# tpm2_pcrread
[...]
sha256:
  0 : 0xD27CC12614B5F4FF85ED109495E320FB1E5495EB28D507E952D51091E7AE2A72
  1 : 0x5BD47F522594A1B6F257426CE5D88CDEDCA5F8D36010F7C37250EE035500B6CC
  2 : 0x1FD6203A96B376B2FADAC2C5AF3C3FE58265F6C86BB510858D2DC838BBCAF6BC
  3 : 0xD27CC12614B5F4FF85ED109495E320FB1E5495EB28D507E952D51091E7AE2A72
  4 : 0x7600BE692F81F6BAAA495180C2389F0E92AF751B423B28DF10822EF031170F33
  5 : 0xE687514A1DE5E9B2370871FBDD7EA1B7482A725B72C12D9F9E04F7844899AF72
  6 : 0xD27CC12614B5F4FF85ED109495E320FB1E5495EB28D507E952D51091E7AE2A72
  7 : 0xD27CC12614B5F4FF85ED109495E320FB1E5495EB28D507E952D51091E7AE2A72
  8 : 0x0000000000000000000000000000000000000000000000000000000000000000
  9 : 0x0000000000000000000000000000000000000000000000000000000000000000
  10: 0x73F386317FA0BD385230E840FBAC09CDD126D9426F1FD26277993ECA5FA35A7E
  11: 0x0000000000000000000000000000000000000000000000000000000000000000
  12: 0x0000000000000000000000000000000000000000000000000000000000000000
  13: 0x0000000000000000000000000000000000000000000000000000000000000000
  14: 0x0000000000000000000000000000000000000000000000000000000000000000
  15: 0x0000000000000000000000000000000000000000000000000000000000000000
  16: 0x0000000000000000000000000000000000000000000000000000000000000000
  17: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  18: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  19: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  20: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  21: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  22: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
  23: 0x0000000000000000000000000000000000000000000000000000000000000000
[...]
```

**Comment by henryreed** [comment]
I made a quick script that converts SWTPM logs to a PCAP, to allow the use of the Wireshark dissector in understanding TPM logs: https://github.com/henryreed/swtpm-log-to-pcap

Would this help with your use case? I am open to any pull requests or suggestions.

**Reply by stefanberger** [reply]
A pull request would be great. Even better would be to support something like `--log format=pcap,...` at some point.

**Reply by henryreed** [reply]
@stefanberger Just so I understand--you want this kind of feature integrated into `swtpm` itself? The script I linked is just Python + Scapy, and requires Scapy to be installed. Did you want something in pure C to keep dependencies down or something that pulls from Python? If it's the former, I'm unfortunately severely constrained at this time, so a pure C rewrite is not something I can support.

**Reply by stefanberger** [reply]
Ok, I will look into it when I have some time. So, yes, I mean it should go directly into swtpm's logging.

**Reply by henryreed** [reply]
Gotcha. This will be tough for two very similar reasons.

1. Even if we add a build option that builds swtpm with Python 3 support and hooks into my script, which is a bit messy, the packet manipulation library I am using is `Scapy`, which is GPL 2.0 only, and would "infect" the `swtpm` project, which is BSD 3 Clause
2. Even if we somehow called on the Wireshark dissector directly, Wireshark and all its dissectors are also GPL 2.0 only, looping back to issue 1

If you wish to have a comprehensive solution within `swtpm` itself, it'll need to rely neither on Scapy nor on a Wireshark dissector. Human-readable logging will need to be built into the project from ground up, or `swtpm` will need to generate PCAPs on its own based on the hexadecimal outputs.

**Reply by stefanberger** [reply]
swtpm would have to implement writing pcap type of packets: https://www.endace.com/learn/what-is-a-pcap-file  
It doesn't look too difficult. It would probably have to also simulate the syn/syn+ack/ack packets that you had to add.
